cpp/src/hash/spark_murmur_hash.cu

```diff
@@ -0,0 +1,183 @@
+/*
+ * Copyright (c) 2019-2022, NVIDIA CORPORATION.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+#include <cudf/column/column_factories.hpp>
+#include <cudf/detail/hashing.hpp>
+#include <cudf/detail/utilities/hash_functions.cuh>
+#include <cudf/detail/utilities/vector_factories.hpp>
+#include <cudf/table/experimental/row_operators.cuh>
+#include <cudf/table/table_device_view.cuh>
+
+#include <rmm/cuda_stream_view.hpp>
+#include <rmm/exec_policy.hpp>
+
+#include <thrust/tabulate.h>
+
+namespace cudf {
+namespace detail {
+
+namespace {
+
+/**
+ * @brief Computes the hash value of a row in the given table.
+ *
+ * This functor uses Spark conventions for MurmurHash3_32 hashing, which
+ * differs from the standard MurmurHash3_32 implementation. These differences
+ * include:
+ * - specialized tail processing in the hash functor
+ * - specializations for decimal types
+ * - special behavior for serially using the output hash as an input seed for
+ *   the next item
+ * - ignorance of null values: [1], [1, null], and [null, 1] have the same hash
+ *
+ * @tparam hash_function Hash functor to use for hashing elements.
+ * @tparam Nullate A cudf::nullate type describing whether to check for nulls.
+ */
+template <template <typename> class hash_function, typename Nullate>
+class spark_device_row_hasher {
```

Comment:
It looks like it would only buy you about a 20% reduction in LOC while potentially adding complexity, so I'm not sure if this is a good idea, but I'll ask just to raise the question: would we benefit from having this inherit from `device_row_hasher<...>`? You'd get some pieces for free (mainly the public API), but you'd still have to create the private hasher adapter.

Reply:
I am not opposed to this in principle but I think I would have a hard time knowing what to change here, to be honest. 😅 I don't know how the complicated `friend` templates and private adapter classes in the class scope would work with inheritance. Guidance would be welcome if you think this is beneficial and could show me how to do it on a call -- this was just the least complex thing I could come up with.

Reply:
I think this is doable but not worth exploring right now. Let's come back to this in 22.10. I'd like to explore the issues around the how we template the row hasher a little more closely first before trying this.